File: trunk/res/html/src/PluginsView.js
/* $Id$
 * IMPOMEZIA Simple Chat
 * Copyright (c) 2008-2012 IMPOMEZIA <schat@impomezia.com>
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

var Plugins = {
  add: function(data)
  {
    var html = '<div class="plugin-wrapper" id="' + data.id + '">' +
                  '<div class="plugin" style="background-image:url(' + data.icon + ');">' +
                    '<div class="plugin-details">' +
                      '<div class="plugin-header"><span class="plugin-title">' + data.title + '</span><span class="plugin-version">' + data.version + '</span></div>' +
                      '<div class="desc-description">' + data.desc + '</div>' +
                    '</div>' +
                    '<div class="plugin-controls">' +
                      '<div class="checkbox enable-checkbox">' +
                        '<label>' +
                          '<input type="checkbox"> ' +
                          '<span class="enable-checkbox-text">' +
                            '<span class="enabled-text" data-tr="enabled">' + Plugins.tr('enabled') + '</span>' +
                            '<span class="enable-text" data-tr="enable">' + Plugins.tr('enable') + '</span>' +
                          '</span>' +
                        '</label>' +
                      '</div>' +
                    '</div>' +
                  '</div>' +
                '</div>';

    $('#plugins-list').append(html);

    Plugins.enable(data.id, data.enabled);

    $('#' + data.id + ' input:checkbox').click(function() {
      Plugins.enable(data.id, $('#' + data.id + ' input:checkbox').prop('checked'));
    });
  },


  enable: function(id, enable)
  {
    if (enable) {
      $('#' + id + ' input:checkbox').attr('checked', 'checked');
      $('#' + id).addClass('plugin-enabled');
      $('#' + id + ' .enabled-text').show();
      $('#' + id + ' .enable-text').hide();
    }
    else {
      $('#' + id + ' input:checkbox').removeAttr('checked');
      $('#' + id).removeClass('plugin-enabled');
      $('#' + id + ' .enabled-text').hide();
      $('#' + id + ' .enable-text').show();
    }
  },

  retranslate: function() {
    $("[data-tr]").each(function() {
      Plugins.TR($(this).attr("data-tr"));
    });
  },

  tr: function(key) {
    return SimpleChat.translate(key);
  },

  TR: function(key) {
    $("[data-tr='" + key + "']").html(Plugins.tr(key));
  }
};

$(document).ready(function() {
  var plugins = PluginsView.list();
  for (var i = 0; i < plugins.length; i++) {
    Plugins.add(plugins[i]);
  }
});

if (typeof PluginsView === "undefined") {
  PluginsView = {
    list: function() { return []; }
  };
}
else {
  SimpleChat.retranslated.connect(Plugins.retranslate);
}

File: src/schat2/plugins/PluginsView.cpp
/* $Id$
 * IMPOMEZIA Simple Chat
 * Copyright © 2008-2012 IMPOMEZIA <schat@impomezia.com>
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include <QApplication>
#include <QFile>
#include <QTimer>
#include <QWebFrame>

#include "ChatCore.h"
#include "ChatNotify.h"
#include "ChatPlugins.h"
#include "ChatSettings.h"
#include "plugins/PluginsView.h"
#include "sglobal.h"
#include "Translation.h"
#include "WebBridge.h"

PluginsView::PluginsView(QWidget *parent)
  : QWebView(parent)
{
  setAcceptDrops(false);
  page()->setLinkDelegationPolicy(QWebPage::DelegateAllLinks);

  connect(page()->mainFrame(), SIGNAL(javaScriptWindowObjectCleared()), SLOT(populateJavaScriptWindowObject()));

  QTimer::singleShot(0, this, SLOT(boot()));
}


/*!
 * Состояние плагина.
 *
 * \return - 0 - Состояние плагина не будет изменено после перезапуска.
 *         - 1 - Плагин будет включен после перезапуска чата.
 *         - 2 - Плагин будет отключен после перезапуска чата.
 */
int PluginsView::state(const QString &id)
{
  PluginItem *item = ChatCore::plugins()->plugin(id);
  if (!item)
    return 0;

  bool enabled = ChatCore::settings()->value(LS("Plugins/") + id).toBool();
  if (item->isLoaded() == enabled)
    return 0;

  if (enabled && !item->isLoaded())
    return 1;

  if (!enabled && item->isLoaded())
    return 2;

  return 0;
}


QVariantList PluginsView::list() const
{
  QVariantList plugins;
  foreach (PluginItem *item, ChatCore::plugins()->list()) {
    QVariantMap data;
    data[LS("id")]      = item->id();
    data[LS("icon")]    = LS("qrc") + item->icon();
    data[LS("title")]   = item->header().value(LS("Name"));
    data[LS("version")] = item->header().value(LS("Version"));
    data[LS("desc")]    = desc(item->header());
    data[LS("enabled")] = ChatCore::settings()->value(LS("Plugins/") + item->id()).toBool();
    plugins.append(data);
  }

  return plugins;
}


void PluginsView::enable(const QString &id, bool enable)
{
  ChatCore::settings()->setValue(LS("Plugins/") + id, enable);
}


void PluginsView::restart()
{
  ChatNotify::start(Notify::Restart);
}


void PluginsView::boot()
{
  QString file = QApplication::applicationDirPath() + LS("/styles/test/html/Plugins.html");
  if (QFile::exists(file))
    file = QUrl::fromLocalFile(file).toString();
  else
    file = LS("qrc:/html/Plugins.html");

  setUrl(QUrl(file));
}


void PluginsView::populateJavaScriptWindowObject()
{
  page()->mainFrame()->addToJavaScriptWindowObject(LS("SimpleChat"), WebBridge::i());
  page()->mainFrame()->addToJavaScriptWindowObject(LS("PluginsView"), this);
}


QString PluginsView::desc(const QVariantMap &data) const
{
  QString desc = data.value(LS("Desc/") + ChatCore::translation()->name()).toString();
  if (!desc.isEmpty())
    return desc;

  return data.value(LS("Desc")).toString();
}
